### ISSUE TITLE
NullPointerException from FlowTypeCheck

### ISSUE THREAD

**Issue by DavePearce** [Whiley/WhileyCompiler#951]
The following causes a `NullPointerException` when it should report a syntax error:

```
type State is { int current }
type Transformer<T> is function(State)->(State)
type Node<T> is { Transformer<T> fn }

method push(State s) -> (State r):
    return s

method f<T>(Transformer<T> fn) -> Node<T>:
    return { fn: fn }

method main() -> Node<State>[]:
    return [f<State>(&(State st -> push(st)))]
```

What's happening is that the type checker is inferring the type of the lambda as a `method` which makes it unsuitable for an argument to `f()`.  Somehow, through all of this jigsaw, a `null` type is returned and saved somewhere causing the exception.  It seems the array initialiser is necessary for this to happen.
